include/targets/generic/kernel_ops.h

```diff
@@ -50,4 +67,109 @@ namespace quda
+     that need tagging.  This can be used as an alternative in cases
+     where the operations are only conditionally used.
+   */
+  struct NoKernelOps {
+    using KernelOpsT = NoKernelOps;
+  };
+
+  /**
+     @brief Used to get KernelOps from a kernel type.  Checks for the
+     existence of KernelOpsT.
+   */
+  template <typename T, typename U = void> struct getKernelOpsS {
+    using type = NoKernelOps;
+  };
+  template <typename T> struct getKernelOpsS<T, std::conditional_t<true, void, typename T::KernelOpsT>> {
+    using type = typename T::KernelOpsT;
+  };
+  template <typename T> using getKernelOps = typename getKernelOpsS<T>::type;
+
+  /**
+     @brief Checks whether a kernel type is tagged with any KernelOps.
+   */
+  template <typename T> static constexpr bool hasKernelOps = !std::is_same_v<getKernelOps<T>, NoKernelOps>;
+
+  /**
+     @brief Checks if first template type matches any of the ops in
+     the second template type, which is a KernelOps template type.
+   */
+  template <typename T, typename U> static constexpr bool hasKernelOp = false;
+  template <typename T, typename... U>
+  static constexpr bool hasKernelOp<T, KernelOps<U...>> = (std::is_same_v<T, U> || ...);
+
+  /**
+     @brief Function to statically check if the passed kernel functor was tagged with all the
+     operations in the template parameters.
+   */
+  template <typename... T, typename Ops> static constexpr void checkKernelOps(const Ops &)
+  {
+    static_assert((hasKernelOp<T, typename Ops::KernelOpsT> || ...));
+  }
+
+  /**
+     @brief Helper to combine two KernelOps or NoKernelOps types.
+   */
+  template <typename... T> struct combineOpsS {
+  };
+  template <typename... T> struct combineOpsS<NoKernelOps, KernelOps<T...>> {
+    using type = KernelOps<T...>;
+  };
+  template <typename... T> struct combineOpsS<KernelOps<T...>, NoKernelOps> {
+    using type = KernelOps<T...>;
+  };
+  template <typename... T, typename... U> struct combineOpsS<KernelOps<T...>, KernelOps<U...>> {
+    using type = KernelOps<T..., U...>;
+  };
+  template <typename T, typename U> using combineOps = typename combineOpsS<T, U>::type;
+
+  /**
+     @brief Gets the total shared memory size needed for a set of
+     kernel operations.  If any ops types have an offset for the
+     shared memory, then the offset is included in the size.
+   */
+  template <typename T> struct sharedMemSizeS {
+    template <typename... Arg> static constexpr unsigned int size(dim3 block, const Arg &...arg)
+    {
+      return T::shared_mem_size(block, arg...);
+    }
+  };
+  template <> struct sharedMemSizeS<NoKernelOps> {
+    template <typename... Arg> static constexpr unsigned int size(dim3, const Arg &...) { return 0; }
+  };
+  template <typename... T> struct sharedMemSizeS<KernelOps<T...>> {
+    template <typename... Arg> static constexpr unsigned int size(dim3 block, const Arg &...arg)
+    {
+      return std::max({sharedMemSizeS<T>::size(block, arg...)...});
```

Comment:
I need to add more documentation, I'll address that in an update.  The short answer is that it is assumed that shared memory is used sequentially (or at least not concurrently), so that two different objects can use the same shared memory, unless the offset has been explicitly specified (in which case the offset has already forced the needed memory size for the offset object to include a sum).  So the total size needed among all objects is just the max of the size of each.

Reply:
Ah yeah, more documentation would help. Your rationale makes sense and is consistent with how you engineered this, I like how it's not going to allocate more shared memory than necessary by construction.

I think my big concern is it'll be hard to debug an issue when someone accidentally forgets to use `combineOps`, etc, to accumulate the total needed shared memory... but between extra documentation and a few of us being familiar with the details, it'll be easy enough to help other developers debug.